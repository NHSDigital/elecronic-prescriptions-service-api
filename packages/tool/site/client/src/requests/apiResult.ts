import {FhirResource} from "fhir/r4"

export function isApiResult(data: unknown): data is ApiResult {
  const result = data as ApiResult
  return typeof result.success === "boolean"
    && "request" in result
    && "response" in result
}

export interface ApiResult {
  success: boolean
  request: FhirResource
  request_xml?: string
  response: FhirResource
  response_xml?: string
}
