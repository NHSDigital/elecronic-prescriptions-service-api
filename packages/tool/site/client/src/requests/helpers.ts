export interface HubResponse {
  certificate: string,
  status_code: number,
  status_string: string,
  message: string,
  timestamp: string,
  signatures: Array<Signature>,
  failed_signatures: Array<Signature>
}

export interface Signature {
  id: string,
  signature: string
}

export function getToken(queryString: URLSearchParams): string {
  const token = queryString.get("token")

  if (!token) {
    throw new TypeError("Token not present in query string.")
  }

  return token
}

export function getState(queryString: URLSearchParams): string | null {
  return queryString.get("state")
}

export function getQueryString(): URLSearchParams {
  const queryString = window.location.search
  if (!queryString) {
    throw new TypeError("Empty query string.")
  }

  return new URLSearchParams(queryString)
}

export async function start(signFn: (jwt: string) => Promise<HubResponse>): Promise<void> {
  try {
    //showStatus("Signing in progress", "Downloading the signature request from the Signing Service.")
    const queryString = getQueryString()
    const token = getToken(queryString)
    const signatureRequest = await retrieveSignatureRequestFromServer(token)

    //showStatus("Signing in progress", "Communicating with Credential Management to generate a signature.")
    const hubResponse = await signFn(signatureRequest)
    validateHubResponse(hubResponse)

    //showStatus("Signing in progress", "Uploading the signature to the Signing Service.")
    const redirectUrl = await sendResponseToProvider(hubResponse, token)

    //showStatus("Signing complete", "Sending you back to the calling application.")
    const state = getState(queryString)
    window.location.href = state ? `${redirectUrl}&state=${state}` : redirectUrl
  } catch (e) {
    console.log(e)
    showError(e.message)
  }
}

function validateHubResponse(hubResponse: HubResponse) {
  if (!hubResponse) {
    throw new TypeError("No response from Credential Management.")
  }

  if (hubResponse.status_code !== 0) {
    console.log(`Got error response from Credential Management: ${JSON.stringify(hubResponse)}`)
    throw new TypeError("Error response from Credential Management.")
  }

  if (!hubResponse.certificate || !hubResponse.signatures?.length) {
    console.log(`Got invalid response from Credential Management: ${JSON.stringify(hubResponse)}`)
    throw new TypeError("Invalid response from Credential Management.")
  }
}

async function retrieveSignatureRequestFromServer(token: string): Promise<string> {
  const response = await fetch(`./client/signaturerequest/${token}`, {
    headers: {
      "X-CSRF-Token": getCSRFTokenFromCookie()
    }
  })
  if (response.status !== 200) {
    console.error(`Got unexpected response from Signing Service: ${response.status}`)
    throw new Error("Failed to retrieve signature request from Signing Service.")
  }
  const responseText = await response.text()
  if (!responseText) {
    throw new Error("Failed to parse response from Signing Service.")
  }
  return responseText
}

function getCSRFTokenFromCookie(): string {
  const cookies = getCookies()
  const crumb = cookies.get("crumb")

  if (!crumb) {
    throw new Error("Could not find CSRF token")
  }

  return crumb
}

function getCookies(): Map<string, string> {
  try {
    return new Map(document.cookie.split(";").map(cookie => {
      const cookieKey = cookie.split("=")[0].trim()
      const cookieValue = cookie.split("=")[1].trim()

      return [cookieKey, cookieValue]
    }))
  } catch(e) {
    console.error("Could not read cookies")
    console.error(e)
    return new Map()
  }
}

// eslint-disable-next-line @typescript-eslint/no-non-null-assertion
const element = (identifier: string) => document.getElementById(identifier)!

export function showError(details?: string): void {
  element("status-container").style.display = "none"
  if (details) {
    element("error-details").innerText = `Error details: ${details}`
  }
  element("error-container").style.display = "block"
}

export async function sendResponseToProvider(hubResponse: HubResponse, token: string): Promise<string> {
  const response = await fetch(
    `./client/signatureresponse/${token}`,
    {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
        "X-CSRF-Token": getCSRFTokenFromCookie()
      },
      body: JSON.stringify({
        "signatures": hubResponse.signatures,
        "certificate": hubResponse.certificate
      })
    }
  )

  if (response.status !== 200) {
    console.error(`Got unexpected response from Signing Service: ${response.status}`)
    throw new Error("Failed to upload signature to Signing Service.")
  }

  const responseJson = await response.json()
  if (!responseJson) {
    throw new Error("Failed to parse response from Signing Service.")
  }

  return responseJson.redirectUri
}
