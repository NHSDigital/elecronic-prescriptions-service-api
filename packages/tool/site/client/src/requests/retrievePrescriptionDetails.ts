/* eslint-disable max-len */
import {axiosInstance} from "./axiosInstance"
import * as fhir from "fhir/r4"
import {getResponseDataIfValid} from "./getValidResponse"
import {getArrayTypeGuard, isBundle} from "../fhir/typeGuards"

export async function getPrescriptionOrderMessage(baseUrl: string, prescriptionId: string): Promise<fhir.Bundle> {
  const prescriptionOrderResponse = await axiosInstance.get<fhir.Bundle>(`${baseUrl}dispense/release/${prescriptionId}`)
  return getResponseDataIfValid(prescriptionOrderResponse, isBundle)
}

export async function getDispenseNotificationMessages(baseUrl: string, prescriptionId: string): Promise<Array<fhir.Bundle>> {
  const dispenseNotificationsResponse = await axiosInstance.get<Array<fhir.Bundle>>(`${baseUrl}dispenseNotifications/${prescriptionId}`)
  return getResponseDataIfValid(dispenseNotificationsResponse, getArrayTypeGuard(isBundle))
}
