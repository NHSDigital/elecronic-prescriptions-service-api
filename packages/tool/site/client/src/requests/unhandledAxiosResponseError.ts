import {AxiosResponse} from "axios"

export class UnhandledAxiosResponseError extends Error {
  response: AxiosResponse

  constructor(message: string, response: AxiosResponse) {
    super(message)
    this.response = response
  }
}
