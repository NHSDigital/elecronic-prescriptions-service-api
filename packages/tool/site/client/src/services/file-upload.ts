import {Dispatch, SetStateAction} from "react"

export function readPrescriptionsFromFiles(
  // eslint-disable-next-line no-undef
  files: FileList,
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  prescriptionFilesUploaded: Array<any>,
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  setPrescriptionFilesUploaded: Dispatch<SetStateAction<Array<any>>>
): void {
  for (let i = 0; i < files.length; i++) {
    // eslint-disable-next-line no-undef
    const reader = new FileReader()
    reader.onload = event => {
      const result = event.target.result
      prescriptionFilesUploaded.push(result)
      setPrescriptionFilesUploaded(prescriptionFilesUploaded)
    }
    reader.readAsText(files[i])
  }
}
