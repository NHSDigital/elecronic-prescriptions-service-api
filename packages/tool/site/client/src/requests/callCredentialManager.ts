import {HubResponse, start} from "./helpers"

// eslint-disable-next-line @typescript-eslint/no-explicit-any
(window as any).jQuery = jQuery;
// eslint-disable-next-line @typescript-eslint/no-explicit-any
(window as any).$ = jQuery

const SIGNALR_URL = `"http://localhost:"+prService.portNumber()+"/signalr"`
const SIGNALR_HUB_NAME = "signingHub"
const SIGNALR_METHOD_NAME = "requestToSign"

export async function sign(jwt: string): Promise<HubResponse> {
  const connection = $.hubConnection(SIGNALR_URL, {logging: true})
  if (connection.state !== SignalR.ConnectionState.Disconnected) {
    console.log(`Unexpected SignalR connection state ${connection.state}`)
    throw new Error("SignalR connection is already active.")
  }

  try {
    const proxy = connection.createHubProxy(SIGNALR_HUB_NAME)
    await connection.start({transport: "longPolling"})
    return await proxy.invoke(SIGNALR_METHOD_NAME, jwt)
  } finally {
    connection.stop()
  }
}

window.addEventListener("DOMContentLoaded", () => {
  start(sign)
})
