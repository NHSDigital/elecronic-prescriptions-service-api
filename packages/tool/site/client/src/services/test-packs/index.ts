import * as XLSX from "xlsx"
import * as fhir from "fhir/r4"
import {getMedicationRequestResources, getMessageHeaderResources} from "../../fhir/bundleResourceFinder"
import {Dispatch, SetStateAction} from "react"
import {createPatient} from "./patients"
import {createPractitioner} from "./practitioners"
import {
  getRowsFromSheet,
  parseOrganisationRowsOrDefault,
  parseAccountRowsOrDefault,
  parsePatientRowsOrDefault,
  parsePrescriptionRows,
  PatientRow,
  PrescriptionRow,
  parsePrescriberRowsOrDefault,
  PrescriberRow,
  OrganisationRow,
  AccountRow
} from "./xls"
import {createPractitionerRole} from "./practitionerRoles"
import {createCommunicationRequest} from "./communicationRequests"
import {createMessageHeader} from "./messageHeader"
import {createPlaceResources} from "./placeResources"
import {createMedicationRequests} from "./medicationRequests"
import {groupBy} from "./helpers"

type PrescriptionData = {
  medicationRows: Array<PrescriptionRow>,
  nominatedPharmacy: string,
  nominatedPharmacyType: string,
  patient: fhir.BundleEntry,
  places: Array<fhir.BundleEntry>,
  practitioner: fhir.BundleEntry,
  practitionerRole: fhir.BundleEntry
}

type PrescriptionCreator = (
  prescriptionData: PrescriptionData,
  prescriptions: Array<fhir.Bundle>
) => void

export const createPrescriptionsFromExcelFile = (
  file: Blob,
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  setPrescriptionsInTestPack: Dispatch<SetStateAction<Array<any>>>,
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  setLoadPageErrors: Dispatch<SetStateAction<any>>
): void => {
  // eslint-disable-next-line no-undef
  const reader = new FileReader()

  reader.onload = function (e) {
    const data = e.target.result
    const workbook = XLSX.read(data, {
      type: "binary"
    })

    const medicationRows = parsePrescriptionRows(getRowsFromSheet("Prescriptions", workbook), setLoadPageErrors)
    const patientRows = parsePatientRowsOrDefault(getRowsFromSheet("Patients", workbook, false), medicationRows.length)
    // eslint-disable-next-line max-len
    const prescriberRows = parsePrescriberRowsOrDefault(getRowsFromSheet("Prescribers", workbook, false), medicationRows.length)
    // eslint-disable-next-line max-len
    const organisationRows = parseOrganisationRowsOrDefault(getRowsFromSheet("Organisations", workbook, false), medicationRows.length)
    const accountRows = parseAccountRowsOrDefault(getRowsFromSheet("Accounts", workbook, false), medicationRows.length)

    setPrescriptionsInTestPack(
      createPrescriptions(
        patientRows,
        prescriberRows,
        organisationRows,
        accountRows,
        medicationRows,
        setLoadPageErrors
      ).map(prescription => JSON.stringify(prescription))
    )
  }

  reader.onerror = function (ex) {
    console.log(ex)
  }

  reader.readAsBinaryString(file)
}

function createPrescriptions(
  patientRows: Map<string, PatientRow>,
  prescriberRows: Map<string, PrescriberRow>,
  organisationRows: Map<string, OrganisationRow>,
  accountRows: Map<string, AccountRow>,
  medicationRows: Array<PrescriptionRow>,
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  setLoadPageErrors: Dispatch<SetStateAction<any>>
): Array<fhir.Bundle> {
  const prescriptions: Array<fhir.Bundle> = []

  const prescriptionRows = groupBy(medicationRows, (row: PrescriptionRow) => row.testId)

  prescriptionRows.forEach((medicationRows, testId) => {
    const accountRow = accountRows.get(testId)
    const organisationRow = organisationRows.get(testId)
    const prescriberRow = prescriberRows.get(testId)
    const prescriptionRow = medicationRows[0]

    const prescriptionData: PrescriptionData = {
      medicationRows: medicationRows,
      nominatedPharmacy: prescriptionRow.nominatedPharmacy,
      nominatedPharmacyType: prescriptionRow.nominatedPharmacyType,
      patient: createPatient(patientRows.get(testId)),
      places: createPlaceResources(organisationRow, accountRow),
      practitioner: createPractitioner(prescriberRow),
      practitionerRole: createPractitionerRole(prescriberRow)
    }

    const prescriptionTreatmentTypeCode = getPrescriptionTreatmentType(prescriptionRow, setLoadPageErrors)

    getCreateFunc(prescriptionTreatmentTypeCode)(prescriptionData, prescriptions)
  })

  return prescriptions
}

function getCreateFunc(prescriptionTreatmentTypeCode: string): PrescriptionCreator {
  switch (prescriptionTreatmentTypeCode) {
    case "acute":
      return createAcutePrescription
    case "continuous":
      return createRepeatPrescribingPrescriptions
    case "continuous-repeat-dispensing":
      return createRepeatDispensingPrescription
    default:
      throw new Error(`Invalid 'Prescription Treatment Type', must be one of: ${validFhirPrescriptionTypes.join(", ")}`)
  }
}

const createAcutePrescription: PrescriptionCreator = (
  prescriptionData: PrescriptionData,
  prescriptions: Array<fhir.Bundle>
) => {
  const prescription = createPrescription(prescriptionData)
  updateNominatedPharmacy(prescription, prescriptionData.nominatedPharmacy)
  updateNominatedPharmacyType(prescription, prescriptionData.nominatedPharmacyType)
  prescriptions.push(prescription)
}

const createRepeatPrescribingPrescriptions: PrescriptionCreator = (
  prescriptionData: PrescriptionData,
  prescriptions: Array<fhir.Bundle>
) => {
  const prescriptionRow = prescriptionData.medicationRows[0]
  const repeatsAllowed = prescriptionRow.repeatsAllowed
  for (let repeatsIssued = 0; repeatsIssued <= repeatsAllowed; repeatsIssued++) {
    const prescription = createPrescription(
      prescriptionData,
      repeatsIssued,
      repeatsAllowed
    )
    updateNominatedPharmacy(prescription, prescriptionData.nominatedPharmacy)
    updateNominatedPharmacyType(prescription, prescriptionData.nominatedPharmacyType)
    prescriptions.push(prescription)
  }
}

const createRepeatDispensingPrescription: PrescriptionCreator = (
  prescriptionData: PrescriptionData,
  prescriptions: Array<fhir.Bundle>
) => {
  const prescriptionRow = prescriptionData.medicationRows[0]
  const prescription = createPrescription(
    prescriptionData,
    0,
    prescriptionRow.repeatsAllowed
  )
  updateNominatedPharmacy(prescription, prescriptionData.nominatedPharmacy)
  updateNominatedPharmacyType(prescription, prescriptionData.nominatedPharmacyType)
  prescriptions.push(prescription)
}

function createPrescription(
  data: PrescriptionData,
  repeatsIssued = 0,
  maxRepeatsAllowed = 0
): fhir.Bundle {

  const prescriptionRow = data.medicationRows[0]

  const fhirPrescription: fhir.Bundle = {
    resourceType: "Bundle",
    id: data.medicationRows[0].testId,
    identifier: {
      system: "https://tools.ietf.org/html/rfc4122",
      value: "ea66ee9d-a981-432f-8c27-6907cbd99219"
    },
    type: "message",
    entry: [
      createMessageHeader(),
      data.patient,
      data.practitioner,
      data.practitionerRole,
      createCommunicationRequest(data.patient, prescriptionRow.additionalInstructions)
    ]
  }

  const resourceType = "Organization"
  const organisation = data.places.map(p => p.resource).find(r => r.resourceType === resourceType) as fhir.Organization
  const odsCode = organisation.identifier.find(i => i.system === "https://fhir.nhs.uk/Id/ods-organization-code").value
  const paddedOdsCode = pad(odsCode, 6)

  createMedicationRequests(
    data.medicationRows,
    paddedOdsCode,
    repeatsIssued,
    maxRepeatsAllowed
  ).forEach(medicationRequest =>
    fhirPrescription.entry.push(medicationRequest)
  )

  fhirPrescription.entry = [
    ...fhirPrescription.entry,
    ...data.places
  ]

  return fhirPrescription
}

function pad(value: string, size: number) {
  value = value.toString()
  while (value.length < size) value = "0" + value
  return value
}

const validFhirPrescriptionTypes = [
  "acute",
  "repeat-dispensing",
  "repeat-prescribing"
]

// eslint-disable-next-line max-len, @typescript-eslint/no-explicit-any
export function getPrescriptionTreatmentType(row: PrescriptionRow, setLoadPageErrors?: Dispatch<SetStateAction<any>>): TreatmentType {
  const code = row.prescriptionTreatmentTypeCode
  if (!validFhirPrescriptionTypes.includes(code)) {
    // eslint-disable-next-line max-len
    const treatmentTypeInvalidError = `Treatment Type column contained an invalid value (${code}). 'Prescription Type' must be one of: ${validFhirPrescriptionTypes.join(", ")}`
    if (setLoadPageErrors) {
      setLoadPageErrors({details: [treatmentTypeInvalidError]})
    }
    throw new Error(treatmentTypeInvalidError)
  }
  switch (code) {
    case "acute":
      return "acute"
    case "repeat-prescribing":
      return "continuous"
    case "repeat-dispensing":
      return "continuous-repeat-dispensing"
  }
}

function updateNominatedPharmacy(bundle: fhir.Bundle, odsCode: string): void {
  if (!odsCode) {
    return
  }
  getMessageHeaderResources(bundle).forEach(messageHeader => {
    messageHeader.destination.forEach(destination => {
      destination.receiver.identifier.value = odsCode
    })
  })
  getMedicationRequestResources(bundle).forEach(function (medicationRequest) {
    medicationRequest.dispenseRequest.performer = {
      identifier: {
        system: "https://fhir.nhs.uk/Id/ods-organization-code",
        value: odsCode
      }
    }
  })
}

function updateNominatedPharmacyType(bundle: fhir.Bundle, performerCode: string): void {
  getMedicationRequestResources(bundle).forEach(function (medicationRequest) {
    medicationRequest.dispenseRequest.extension[0].valueCoding.code = performerCode
  })
}

export type TreatmentType = "acute" | "continuous" | "continuous-repeat-dispensing"
