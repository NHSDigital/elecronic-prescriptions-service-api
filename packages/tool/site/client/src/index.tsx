import * as React from "react"
import {Page} from "./components/page"
import SignPage from "./pages/signPage"
import * as ReactDOM from "react-dom"
import {BrowserRouter, Route, Switch} from "react-router-dom"
import ClaimPage from "./pages/claimPage"
import DispensePage from "./pages/dispensePage"
import {CookiesProvider} from "react-cookie"
import SendPage from "./pages/sendPage"
import HomePage from "./pages/homePage"
import PrescriptionSearchPage from "./pages/prescriptionSearchPage"
import ReleasePage from "./pages/releasePage"
import CancelPage from "./pages/cancelPage"
import MyPrescriptionsPage from "./pages/myPrescriptionsPage"
import LoginPage from "./pages/loginPage"
import ValidatePage from "./pages/validatePage"
import ReturnPage from "./pages/returnPage"
import WithdrawPage from "./pages/withdrawPage"
import LoadPage from "./pages/loadPage"
import {Environment} from "./services/environment"
import LogoutPage from "./pages/logoutPage"
import ConfigPage from "./pages/configPage"
import ComparePage from "./pages/comparePage"
import ViewPrescriptionPage from "./pages/viewPrescriptionPage"
import {PageFullWidth} from "./components/pageFullWidth"
import {PageContainer} from "./components/pageContainer"
import DoseToTextPage from "./pages/doseToTextPage"
import TrackerViewPrescriptionPage from "./pages/trackerViewPrescription"

const customWindow = window as Record<string, any>

export interface AppContextValue {
  baseUrl: string
  environment: Environment
}

export const AppContext = React.createContext<AppContextValue>({baseUrl: "/", environment: "int"})

async function startApplication(baseUrl: string, environment: Environment): Promise<void> {
  const urlParams = new URLSearchParams(window.location.search)
  const content = (
    <AppContext.Provider value={{baseUrl, environment}}>
      <CookiesProvider>
        <Page>
          <BrowserRouter>
            <Switch>
              <Route path={`${baseUrl}/`}>
                <PageContainer>
                  <HomePage />
                </PageContainer>
              </Route>
              <Route path={`${baseUrl}dose-to-text`}>
                <PageContainer>
                  <DoseToTextPage/>
                </PageContainer>
              </Route>
              <Route path={`${baseUrl}config`}>
                <PageContainer>
                  <ConfigPage/>
                </PageContainer>
              </Route>
              <Route path={`${baseUrl}my-prescriptions`}>
                <PageContainer>
                  <MyPrescriptionsPage/>
                </PageContainer>
              </Route>
              <Route path={`${baseUrl}login`}>
                <PageContainer>
                  <LoginPage separateAuth={urlParams.get("separate_auth")}/>
                </PageContainer>
              </Route>
              <Route path={`${baseUrl}logout`}>
                <PageContainer>
                  <LogoutPage/>
                </PageContainer>
              </Route>
              <Route path={`${baseUrl}prescribe/load`}>
                <PageContainer>
                  <LoadPage/>
                </PageContainer>
              </Route>
              <Route path={`${baseUrl}prescribe/edit`}>
                <PageContainer>
                  <SignPage/>
                </PageContainer>
              </Route>
              <Route path={`${baseUrl}prescribe/send`}>
                <PageContainer>
                  <SendPage token={urlParams.get("token")} state={urlParams.get("state")}/>
                </PageContainer>
              </Route>
              <Route path={`${baseUrl}prescribe/cancel`}>
                <PageContainer>
                  <CancelPage prescriptionId={urlParams.get("prescription_id")}/>
                </PageContainer>
              </Route>
              <Route path={`${baseUrl}dispense/release`}>
                <PageContainer>
                  <ReleasePage prescriptionId={urlParams.get("prescription_id")}/>
                </PageContainer>
              </Route>
              <Route path={`${baseUrl}dispense/return`}>
                <PageContainer>
                  <ReturnPage prescriptionId={urlParams.get("prescription_id")} />
                </PageContainer>
              </Route>
              <Route path={`${baseUrl}dispense/withdraw`}>
                <PageContainer>
                  <WithdrawPage prescriptionId={urlParams.get("prescription_id")} />
                </PageContainer>
              </Route>
              <Route path={`${baseUrl}dispense/dispense`}>
                <PageContainer>
                  <DispensePage prescriptionId={urlParams.get("prescription_id")} amendId={urlParams.get("amend_id")} />
                </PageContainer>
              </Route>
              <Route path={`${baseUrl}dispense/claim`}>
                <PageContainer>
                  <ClaimPage prescriptionId={urlParams.get("prescription_id")} amend={!!urlParams.get("amend")}/>
                </PageContainer>
              </Route>
              <Route path={`${baseUrl}search`}>
                <PageContainer>
                  <PrescriptionSearchPage prescriptionId={urlParams.get("prescription_id")} />
                </PageContainer>
              </Route>
              <Route path={`${baseUrl}tracker`}>
                <PageContainer>
                  <TrackerViewPrescriptionPage prescriptionId={urlParams.get("prescription_id")} />
                </PageContainer>
              </Route>
              <Route path={`${baseUrl}view`}>
                <PageContainer>
                  <ViewPrescriptionPage prescriptionId={urlParams.get("prescription_id")} />
                </PageContainer>
              </Route>
              <Route path={`${baseUrl}validate`}>
                <PageContainer>
                  <ValidatePage />
                </PageContainer>
              </Route>
              <Route path={`${baseUrl}compare-prescriptions`}>
                <PageFullWidth>
                  <ComparePage />
                </PageFullWidth>
              </Route>
            </Switch>
          </BrowserRouter>
        </Page>
      </CookiesProvider>
    </AppContext.Provider>
  )
  ReactDOM.render(content, document.getElementById("root"))
}

customWindow.startApplication = startApplication
