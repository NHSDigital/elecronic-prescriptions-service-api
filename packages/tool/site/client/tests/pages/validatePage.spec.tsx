import {waitFor} from "@testing-library/react"
import {fireEvent, screen} from "@testing-library/dom"
import pretty from "pretty"
import * as React from "react"
import MockAdapter from "axios-mock-adapter"
import {AppContextValue} from "../../src"
import {renderWithContext} from "../renderWithContext"
import userEvent from "@testing-library/user-event"
import {axiosInstance} from "../../src/requests/axiosInstance"
import ValidatePage from "../../src/pages/validatePage"
import {internalDev} from "../../src/services/environment"

const baseUrl = "baseUrl/"
const context: AppContextValue = {baseUrl, environment: internalDev}

const validateUrl = `${baseUrl}validate`

const mock = new MockAdapter(axiosInstance)

beforeEach(() => mock.reset())
afterEach(() => mock.reset())

test("Displays validate form", async () => {
  const container = await renderPage()

  expect(screen.getByText("Validate a FHIR Resource")).toBeTruthy()
  expect(pretty(container.innerHTML)).toMatchSnapshot()
})

test("Displays validate result", async () => {
  mock.onAny(validateUrl).reply(200, {
    success: true,
    request: {req: "JSON Request"},
    response: {res: "JSON Response"}
  })

  const container = await renderPage()
  const textArea = container.querySelector("[name='validatePayload']")
  fireEvent.change(textArea, {target: {value: "{}"}})
  userEvent.click(screen.getByText("Validate"))
  await waitFor(() => screen.getByText("Sending validation request."))
  await waitFor(() => screen.getByText(/Validate Result/))
  expect(screen.getByText((/JSON Request/))).toBeTruthy()
  expect(screen.getByText((/JSON Response/))).toBeTruthy()
  expect(pretty(container.innerHTML)).toMatchSnapshot()
})

async function renderPage() {
  const {container} = renderWithContext(<ValidatePage/>, context)
  await waitFor(() => screen.getByText("Validate a FHIR Resource"))
  return container
}
