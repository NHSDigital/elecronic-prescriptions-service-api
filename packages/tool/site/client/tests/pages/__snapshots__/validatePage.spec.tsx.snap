// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Displays validate form 1`] = `
"<h1 class="nhsuk-label-wrapper"><label class="nhsuk-label nhsuk-label--xl">Validate a FHIR Resource</label></h1>
<form>
  <div class="nhsuk-form-group">
    <fieldset class="nhsuk-fieldset sc-cwHqhk kMCanE">
      <div class="nhsuk-form-group"><textarea class="nhsuk-textarea" name="validatePayload" id="validatePayload" rows="20"></textarea></div>
    </fieldset>
  </div>
  <div class="sc-eqUzNf chXQhu"><button class="nhsuk-button" aria-disabled="false" type="submit">Validate</button><button class="nhsuk-button nhsuk-button--secondary" aria-disabled="false" type="button">Back</button></div>
</form>"
`;

exports[`Displays validate result 1`] = `
"<h1 class="nhsuk-label-wrapper"><label class="nhsuk-label nhsuk-label--xl">Validate Result <svg class="nhsuk-icon nhsuk-icon__tick" xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" aria-hidden="true" height="34" width="34">
      <path stroke-width="4" stroke-linecap="round" stroke="#007f3b" fill="none" d="M18.4 7.8l-8.5 8.4L5.6 12"></path>
    </svg></label></h1>
<div class="nhsuk-expander-group">
  <details class="nhsuk-details nhsuk-expander">
    <summary class="nhsuk-details__summary"><span class="nhsuk-details__summary-text">1 - FHIR Request that a supplier would send to the FHIR API</span></summary>
    <div class="nhsuk-details__text">
      <div class="sc-eqUzNf chXQhu"><button class="nhsuk-button sc-iGgVNO cEsNDy" aria-disabled="false" type="submit">Copy</button><a class="nhsuk-button sc-iGgVNO cEsNDy" role="button" aria-disabled="false" draggable="false" href="data:application/json;charset=utf-8,%7B%0A%20%20%22req%22%3A%20%22JSON%20Request%22%0A%7D" download="message">Download</a></div>
      <div class="_2IvMF _GzYRV">
        <div class="_2bkNM" role="list"><span class="_oLqym _f10Tu _1MFti _1LId0" role="button" tabindex="0" aria-label="collapse JSON" aria-expanded="true" aria-controls=":jsnvw:1:"></span><span class="_3uHL6 _3eOF8">{</span>
          <div id=":jsnvw:1:">
            <div class="_2bkNM" role="listitem"><span class="_1MGIk">req:</span><span class="_vGjyY">"JSON Request"</span></div>
          </div><span class="_3uHL6 _3eOF8">}</span>
        </div>
      </div>
    </div>
  </details>
  <details class="nhsuk-details nhsuk-expander">
    <summary class="nhsuk-details__summary"><span class="nhsuk-details__summary-text">2 - FHIR Response that a supplier will receive from the FHIR API</span></summary>
    <div class="nhsuk-details__text">
      <div class="sc-eqUzNf chXQhu"><button class="nhsuk-button sc-iGgVNO cEsNDy" aria-disabled="false" type="submit">Copy</button><a class="nhsuk-button sc-iGgVNO cEsNDy" role="button" aria-disabled="false" draggable="false" href="data:application/json;charset=utf-8,%7B%0A%20%20%22res%22%3A%20%22JSON%20Response%22%0A%7D" download="message">Download</a></div>
      <div class="_2IvMF _GzYRV">
        <div class="_2bkNM" role="list"><span class="_oLqym _f10Tu _1MFti _1LId0" role="button" tabindex="0" aria-label="collapse JSON" aria-expanded="true" aria-controls=":jsnvw:2:"></span><span class="_3uHL6 _3eOF8">{</span>
          <div id=":jsnvw:2:">
            <div class="_2bkNM" role="listitem"><span class="_1MGIk">res:</span><span class="_vGjyY">"JSON Response"</span></div>
          </div><span class="_3uHL6 _3eOF8">}</span>
        </div>
      </div>
    </div>
  </details>
</div>
<div class="sc-eqUzNf chXQhu"><button class="nhsuk-button nhsuk-button--secondary" aria-disabled="false" type="button">Back</button></div>"
`;
